partiql-lang/src/main/kotlin/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -2159,6 +2206,173 @@ internal open class EvaluatingCompiler(
             CompiledGroupByItem(alias.text.exprValue(), uniqueName, compileAstExpr(it.expr))
         }
 
+    /**
+     * Represents all the compiled `EXCLUDE` paths that start with the same [CompiledExcludeExpr.root]. Notably,
+     * redundant paths (i.e. exclude paths that exclude values already excluded by other paths) will be removed.
+     */
+    internal data class CompiledExcludeExpr(val root: PartiqlAst.Identifier, val exclusions: RemoveAndOtherSteps)
```

Comment:
From offline discussion, some of this PR's code in `EvaluatingCompiler` can be reused when porting `EXCLUDE` to the physical plan compiler. Some logic here will be refactored in an upcoming PR that adds `EXCLUDE` to the other compiler.